### PR TITLE
generate command 

### PR DESCRIPTION
**Pull request by pelikhan**
Implement PromptPex strategy to generate tests for prompts automatically.

<hr/>

# 🚀 PromptPex Power-Up: Smarter Prompt Test Generation, Output Cleaning, and Dev Experience Improvements

This PR supercharges the `gh models` CLI extension with a suite of new features and quality-of-life upgrades focused on prompt test generation, output handling, and developer workflow. Highlights include:

- 🧪 **PromptPex-Based Test Generation**  
  - New `generate` command leveraging the PromptPex framework for systematic, rules-driven prompt test creation.
  - Extensive documentation and examples to guide users through advanced prompt testing workflows.

- 🧹 **Model Output Cleaning Utilities**  
  - Added helpers to normalize and sanitize LLM-generated outputs, removing formatting artifacts and detecting refusal responses.
  - Comprehensive unit tests ensure robust text cleaning.

- 🧠 **Context Persistence & Effort Configuration**  
  - Support for loading, saving, and merging prompt test contexts, enabling session persistence and collaborative editing.
  - Configurable "effort levels" to control the intensity and scope of test generation runs.

- 🏆 **Rules-Based Output Evaluation**  
  - Integrated rules compliance evaluator for automated, granular assessment of LLM outputs against prompt rules.

- 🔬 **Advanced Parsing & Robustness**  
  - Improved JSON extraction and normalization from LLM responses, handling markdown code blocks, JS-style concatenations, and malformed content.
  - Flexible rules parsing from varied list formats.

- 🛠️ **Utility Functions & CLI Enhancements**  
  - New helpers for string slicing, map merging, and template variable parsing.
  - Enhanced CLI output formatting and verbose logging for better feedback.

- 🧪 **Expanded Unit Test Coverage**  
  - Table-driven and scenario-based tests for flag parsing, context management, error handling, and end-to-end prompt rendering.

- 📝 **Developer Experience & CI Improvements**  
  - Updated Makefile with new linting targets and improved build workflows.
  - Rich project and AI agent instructions for onboarding and code consistency.

- 🗂️ **Prompt File Handling Upgrades**  
  - Cleaner YAML outputs with `omitempty` struct tags.
  - Round-trip file save support and improved type safety for test data.

These changes collectively deliver smarter, more reliable prompt test automation, improved output handling, and a better developer experience for working with LLM prompts and tests.

> AI-generated content by prd may be incorrect.

### CODE REVIEW

File: cmd/generate/export.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package generate
+
+/*
```
Comment (sgoedecke) [inline cmd/generate/export.go:3]: Should this be uncommented or should the file be removed?

File: genaisrc/prd.genai.mts [hunk 2]
```diff
@@ -0,0 +1,63 @@
+script({
```
Comment (sgoedecke) [inline genaisrc/prd.genai.mts:1]: Do we need to keep the genaisrc/ folder?

File: pkg/util/util.go [hunk 3]
```diff
@@ -18,3 +21,40 @@ func WriteToOut(out io.Writer, message string) {
 func Ptr[T any](value T) *T {
 	return &value
 }
+
+// parseTemplateVariables parses template variables from the --var flags
+func ParseTemplateVariables(flags *pflag.FlagSet) (map[string]string, error) {
+	varFlags, err := flags.GetStringArray("var")
```
Comment (sgoedecke) [inline pkg/util/util.go:27]: Thanks for pulling this bugfix in

**Review by sgoedecke — COMMENTED** [review]
I think this is looking great! Just some initial comments about files we may not need

File: .github/copilot-instructions.md
Comment (cheshire137) [inline .github/copilot-instructions.md:1]: I could see this file being useful for humans to read, too, to get an overview of the repo. Might be valuable to link from the readme. 🤔

File: README.md [hunk 4]
```diff
@@ -84,6 +84,73 @@ Here's a sample GitHub Action that uses the `eval` command to automatically run
...
+```
+
+The `generate` command analyzes your prompt file and automatically creates test cases to evaluate the prompt's behavior across different scenarios and edge cases. This helps ensure your prompts are robust and perform as expected.
+
+##### Advanced options
+
+You can customize the test generation process with various options:
+
+```shell
+# Specify effort level (low, medium, high)
```
Comment (cheshire137) [inline README.md:101]: Can you give any detail about why one would choose a particular effort level? Like is effort level high going to result in the most thorough tests but potentially take a long time?

File: README.md [hunk 5]
```diff
@@ -84,6 +84,73 @@ Here's a sample GitHub Action that uses the `eval` command to automatically run
...
+gh models generate --effort high my_prompt.prompt.yml
+
+# Use a specific model for groundtruth generation
```
Comment (cheshire137) [inline README.md:104]: I haven't seen the term groundtruth before. Is there any good resource we could include a link to, to explain the concept?

File: README.md [hunk 6]
```diff
@@ -84,6 +84,73 @@ Here's a sample GitHub Action that uses the `eval` command to automatically run
...
+gh models generate --groundtruth-model "openai/gpt-4.1" my_prompt.prompt.yml
+
+# Disable groundtruth generation
+gh models generate --groundtruth-model "none" my_prompt.prompt.yml
+
+# Load from existing session file
```
Comment (cheshire137) [inline README.md:110]: How would a session file be created?

File: README.md [hunk 7]
```diff
@@ -84,6 +84,73 @@ Here's a sample GitHub Action that uses the `eval` command to automatically run
...
+```
+
+The command supports custom instructions for different phases of test generation:
+- `--instruction-intent`: Custom system instruction for intent generation
+- `--instruction-inputspec`: Custom system instruction for input specification generation  
+- `--instruction-outputrules`: Custom system instruction for output rules generation
+- `--instruction-inverseoutputrules`: Custom system instruction for inverse output rules generation
+- `--instruction-tests`: Custom system instruction for tests generation
+
+##### Understanding PromptPex
```
Comment (cheshire137) [inline README.md:124]: I could see moving this section earlier, so the definitions for intent analysis, input specification, etc phases appear before the custom instruction parameters for them like `--instruction-intent`.

File: genaisrc/prd.genai.mts (hunk 2 above)
Comment (pelikhan) [reply genaisrc/prd.genai.mts:1]: @copilot remove genaisrc folder and files, remove prr task in make file

File: cmd/generate/evaluators.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	"github.com/github/gh-models/pkg/prompt"
+)
+
+// generateRulesEvaluatorSystemPrompt generates the system prompt for rules evaluation
+func (h *generateCommandHandler) GenerateRulesEvaluator(context *PromptPexContext) prompt.Evaluator {
+	// Get the original prompt content
+	promptContent := RenderMessagesToString(context.Prompt.Messages)
+	rulesContent := strings.Join(context.Rules, "\n")
+
+	systemPrompt := fmt.Sprintf(`Your task is to very carefully and thoroughly evaluate the given output generated by a chatbot in <chatbot_output> to find out if it comply with its prompt and the output rules that are extracted from the description and provided to you in <output_rules>.
```
[the remark this answers is not in the document]
Comment (pelikhan) [reply cmd/generate/evaluators.go:16]: I'm not exactly sure how we have resource files in Go but yes much easier to keep the prompts in separate file.

**Review by pelikhan — COMMENTED** [review]
@copilot Analyze comments around documentation and apply fixes.

**Comment by pelikhan** [comment]
@copilot address comments

File: cmd/generate/cleaner_test.go [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+			expected: true,
+		},
+		{
+			name:     "detects 'i'm sorry' mixed case",
+			input:    "I'm Sorry, But I Cannot Help",
+			expected: true,
+		},
+		{
+			name:     "detects phrase within larger text",
+			input:    "Unfortunately, I can't assist with that particular request. Please try something else.",
```
[the remark this answers is not in the document]
Comment (pelikhan) [reply cmd/generate/cleaner_test.go:37]: LLMs are flimsy and somethings the promptpex prompts referencing inlined prompts confuses the inference engine.

File: cmd/generate/pipeline.go [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+	return nil
+}
+
+// generateInverseRules generates inverse rules (what makes an invalid output)
+func (h *generateCommandHandler) generateInverseRules(context *PromptPexContext) error {
+	h.WriteStartBox("Inverse output rules", "")
+	if len(context.InverseRules) == 0 {
+
+		system := `Based on the following <output_rules>, generate inverse rules that describe what would make an INVALID output.
+These should be the opposite or negation of the original rules.`
```
[the remark this answers is not in the document]
Comment (pelikhan) [reply cmd/generate/pipeline.go:237]: do you mean an example of prompt/rules/inverse rules in the readme?

**Comment by pelikhan** [comment]
I address a few comments.

File: cmd/generate/evaluators.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sgoedecke) [reply cmd/generate/evaluators.go:16]: Compiling a prompt into the Golang binary is an interesting idea. We probably want something like this to support prompt.yml in Golang programs.

The best option would be to use Golang's `embed`, I think, with our existing prompt.yml File [type](https://github.com/github/gh-models/blob/main/pkg/prompt/prompt.go#L15):

```
import (
    "embed"
    "gopkg.in/yaml.v3"
)

//go:embed config.yml
var configYAML []byte

func main() {
    var data prompt.File
    yaml.Unmarshal(configYAML, &data)
    // use 'data'...
}
```

I'm happy for you to do this in a followup after merging this PR, though, given the size.

**Review by sgoedecke — APPROVED** [review]
Happy to approve again after merge conflicts are fixed

**Comment by pelikhan** [comment]
@sgoedecke I added a 'min' effort level that aggressively disables the amount of tests generated.

**Review by sgoedecke — APPROVED** [review]

**Review by sgoedecke — APPROVED** [review]
